core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinJoinRemoveRule.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.RelFactories;
+import org.apache.calcite.rel.logical.LogicalAggregate;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that matches an {@link org.apache.calcite.rel.core.Aggregate}
+ * on a {@link org.apache.calcite.rel.core.Join} and removes the left input
+ * of the join provided that the left input is also a left join if possible.
+ *
+ * <p>For instance,</p>
+ *
+ * <blockquote>
+ * <pre>select distinct s.product_id, pc.product_id from
+ * sales as s
+ * left join product as p
+ * on s.product_id = p.product_id
+ * left join product_class pc
+ * on s.product_id = pc.product_id</pre></blockquote>
+ *
+ * <p>becomes
+ *
+ * <blockquote>
+ * <pre>select distinct s.product_id, pc.product_id from
+ * sales as s
+ * left join product_class pc
+ * on s.product_id = pc.product_id</pre></blockquote>
+ *
+ */
+public class AggregateJoinJoinRemoveRule extends RelOptRule {
+  public static final AggregateJoinJoinRemoveRule INSTANCE
+      = new AggregateJoinJoinRemoveRule(LogicalAggregate.class,
+      LogicalJoin.class, RelFactories.LOGICAL_BUILDER);
+
+  /** Creates an AggregateJoinJoinRemoveRule. */
+  public AggregateJoinJoinRemoveRule(
+      Class<? extends Aggregate> aggregateClass,
+      Class<? extends Join> joinClass, RelBuilderFactory relBuilderFactory) {
+    super(
+        operand(aggregateClass,
+            operandJ(joinClass, null,
+                join -> join.getJoinType() == JoinRelType.LEFT,
+                operandJ(joinClass, null,
+                    join -> join.getJoinType() == JoinRelType.LEFT, any()))),
+        relBuilderFactory, null);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Aggregate aggregate = call.rel(0);
+    final Join topJoin = call.rel(1);
+    final Join bottomJoin = call.rel(2);
+    int leftBottomChildSize = bottomJoin.getLeft().getRowType()
+        .getFieldCount();
+
+    // Check whether the aggregate uses unexpected columns.
+    if (!RelOptUtil.removeJoin(aggregate, leftBottomChildSize,
+        bottomJoin.getRowType().getFieldCount())) {
+      return;
+    }
+
+    // Check whether the top join uses unexpected columns.
+    final List<Integer> leftKeys = new ArrayList<>();
+    RelOptUtil.splitJoinCondition(topJoin.getLeft(), topJoin.getRight(),
+        topJoin.getCondition(), leftKeys, new ArrayList<>(),
+        new ArrayList<>());
+    if (leftKeys.stream().filter(s -> s >= leftBottomChildSize).count() > 0) {
```

Comment:
I hope the if condition part can be splitted into 2 statements.

Reply:
The `if` condition can be simplified to `leftKeys.stream().anyMatch(s -> s >= leftBottomChildSize)`.